File: 2025-10/2025-10-05/exoplanet_search.py
def has_exoplanet(readings):
    import re
    CODE_OF_UPPERCASE_A = ord('A')
    LUMINOSITY_THRESHOLD = 10 #luminosity level beyond which letters are used to represent values

    luminosity_levels = list(map(lambda reading: 
        ord(reading) - CODE_OF_UPPERCASE_A + LUMINOSITY_THRESHOLD if re.match(r"[A-Z]", reading) 
        else int(reading), 
    readings))

    average_luminosity_level = sum(luminosity_levels) / len(luminosity_levels)

    return any(luminosity_level <= average_luminosity_level * 80 / 100 for luminosity_level in luminosity_levels)

print(has_exoplanet("665544554"))
print(has_exoplanet("FGFFCFFGG"))
print(has_exoplanet("MONOPLONOMONPLNOMPNOMP"))
print(has_exoplanet("FREECODECAMP"))
print(has_exoplanet("9AB98AB9BC98A"))
print(has_exoplanet("ZXXWYZXYWYXZEGZXWYZXYGEE"))

# --- TEST SUITE ---

tests_text = '''
1. has_exoplanet("665544554") should return False.
2. has_exoplanet("FGFFCFFGG") should return True.
3. has_exoplanet("MONOPLONOMONPLNOMPNOMP") should return False.
4. has_exoplanet("FREECODECAMP") should return True.
5. has_exoplanet("9AB98AB9BC98A") should return False.
6. has_exoplanet("ZXXWYZXYWYXZEGZXWYZXYGEE") should return True.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 2025-10/2025-10-05/exoplanetSearch.js
function hasExoplanet(readings) {
  const codeOfUppercaseA = 'A'.charCodeAt(0);
  const luminosityThreshold = 10; //luminosity level beyond which letters are used to represent values

  const luminosityLevels = 
    readings
    .split("")
    .map(reading => /[A-Z]/.exec(reading) ? reading.charCodeAt(0) - codeOfUppercaseA + luminosityThreshold : parseInt(reading))

  const averageLuminosityLevel = 
    luminosityLevels
    .reduce((sum, luminosityLevel) => sum + luminosityLevel)
    / luminosityLevels.length

  return luminosityLevels
    .some(luminosityLevel => luminosityLevel <= averageLuminosityLevel * 80 / 100);
}

console.log(hasExoplanet("665544554"));
console.log(hasExoplanet("FGFFCFFGG"));
console.log(hasExoplanet("MONOPLONOMONPLNOMPNOMP"));
console.log(hasExoplanet("FREECODECAMP"));
console.log(hasExoplanet("9AB98AB9BC98A"));
console.log(hasExoplanet("ZXXWYZXYWYXZEGZXWYZXYGEE"));

// --- TEST SUITE ---

const testsText = `
1. hasExoplanet("665544554") should return false.
2. hasExoplanet("FGFFCFFGG") should return true.
3. hasExoplanet("MONOPLONOMONPLNOMPNOMP") should return false.
4. hasExoplanet("FREECODECAMP") should return true.
5. hasExoplanet("9AB98AB9BC98A") should return false.
6. hasExoplanet("ZXXWYZXYWYXZEGZXWYZXYGEE") should return true.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);